Feat/tokens (#833)
johnsonandjohnson/Bodiless-JS: packages/fclasses/README.md

re-extend their look and feel, with less danger of breakage when the underlying
component changes.

## Tokens

In general, we express a design token as a React higer-order component (HOC).
Applying the HOC to a component is equivalent to styling that component
with a token:

```js
const ComponentWithStyles = withMyStyles(Component);
```

This pattern should be familiar to those who have worked with CSS-in-JS libraries
like Styled Components or Emotion.

Any HOC can be used as a token, and tokens can be composed using normal
functional programming paradigms (eg Lodash flow):
```js
const withComposedToken = flow(
  withToken1,
  withToken2,
);
```

> HoweverBodiless provides a token composition utility which adds some

Author: ```suggestion
However, Bodiless provides a token composition utility which adds some
```

re-extend their look and feel, with less danger of breakage when the underlying
component changes.

## Tokens

> In bodiless, you implement design tokens as React higer-order components (HOC).

Reviewer: bodiless => Bodiless

Author: ```suggestion
In Bodiless, you implement design tokens as React higher-order components (HOC).
```

Applying the HOC to a component is equivalent to styling that component with a
token:

```js
const ComponentWithStyles = withMyStyles(Component);
```

This pattern should be familiar to those who have worked with CSS-in-JS libraries

> like Styled Components or Emotion.

Reviewer: Emotion => [Emotion](https://emotion.sh/docs/introduction)

Author: ```suggestion
like [Styled Components](https://styled-components.com/) or [Emotion](https://emotion.sh/docs/introduction).
```

Any HOC can be used as a token, and tokens can be composed using normal
functional programming paradigms (eg Lodash flow):
```js
const withComposedToken = flow(
  withToken1,
  withToken2,
);
```

However, Bodiless provides a token composition utility which adds some
additional functionality:

- The ability to attach metadata to a token.
- The ability to selectively remove tokens from a composition based on
  their metadata (or other criteria).

This is intended to promote design-system thinking when defining
tokens, by encouraging us to think about the structure and organization
of tokens as we implement them.  It also facilitates implementation of

> tools which allow browsing the design system (eg StorybooK), and eases

Reviewer: StorybooK => Storybook

Author: ```suggestion
tools which allow browsing the design system (eg Storybook), and eases
```

the process of extending or customizing composed tokens without fully
recomposing them.

In general, you can use `asToken` to compose tokens the same way you
would use Lodash flow, eg:

```js
const withComposedToken = asToken(
  withToken1,
  withToken2,
);
```

However, there are a few key differences:

> - Metadata (static properties) attached to a component ar prpagated through

Reviewer 2: ar --> are

Reviewer: prpagated  => propagated

Author: ```suggestion
- Metadata (static properties) attached to a component are prppagated through
```

  the chain of HOC's.
- If you are using Typescript, the type of the parameters is constrained to be an

>   HOC, except...

Reviewer: "except..."? missing exception details?

Author: ```suggestion
  HOC (or an object specifying metadata, see below).
```

- There is an optional overload to accept a "TokenMeta" object which consists of
  metadata which should be attached to the token.
- We intruduce a special kind of Token known as a "Filter". See more
  below.

### Metadata and Filters

Token metadata are properties which can be attached to tokens to help
organize them and understand their structure. When a token is applied,
its metadata will also be attached to the component to which it is applied.
If a composed token is applied, metadata from all constituents will be
aggregated and attached to the target component. See below for some examples.

In addition to a normal HOC, a Token can also be a "filter". A filter is a token
which, when composed with other tokens, *removes* any which match certain
criteria. Filters are usually defined to test the metadata attached to other
tokens. So, for exmple, you can compose a token which removes all 'Color' tokens
and adds a new one.

> Note that while metadata from all constituent tokens are aggregated and attached

> > to the component to which a composed token is applied, the tomposed token

Reviewer 2: tomposed --> composed

Author: ```suggestion
> to the component to which a composed token is applied, the composed token
```

> much harder to filter. Think of the metadata attached to a Token as that portion
> of the final metadata which it will contribute.
>
> It's easy enough to get the aggregated metadata, eg:
> ```
> const finalMeta = pick(myToken(Fragment), 'categories', 'title', ...);
> ```

### Examples

Given

```js
const asBold = asToken(
  addClasses('font-bold'),
  { categories: { Style: ['Bold'] } },
);

const asTextBlue = asToken(
  addClasses('text-blue-500'),
  { categories: { TextColor: ['Blue'] } },
);

const asTextRed = asToken(
  addClasses('text-red-500'),
  { categories: { TextColor: ['Red'] } },
);
// Same as:
// const asTextRed = asToken(addClasses('text-red-500'));
// asTextRed.meta = { categories: { TextColor: ['Red'] } };

const asBgYellow = asToken(
  addClasses('bg-yellow-500'),
  { categories: { BgColor: ['Yellow'] } },
)

const asHeader1 = asToken(
  asTextBlue,
  asBold,
  asBgYellow,
  { categories: { Header: ['H1'] } },
);

const Header1 = asHeader1(H1);  // `H1` is a version of 'h1' stylable with fclasses, see below.
```

Then

```js

<Header1 /> === <h1 className="text-blue bg-yellow-500 font-bold" />

> // The component itself includes aggreaged metadata from all composed tokens...

Reviewer: aggreaged  => aggregated

Author: ```suggestion
// The component itself includes aggregated metadata from all composed tokens...
```
